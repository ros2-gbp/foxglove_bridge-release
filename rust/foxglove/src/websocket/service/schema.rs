use crate::Schema;

/// A service request or response schema.
#[derive(Debug, Clone)]
pub(crate) struct MessageSchema {
    pub encoding: String,
    pub schema: Schema,
}

/// A service schema.
#[derive(Debug, Clone)]
pub struct ServiceSchema {
    name: String,
    request: Option<MessageSchema>,
    response: Option<MessageSchema>,
}
impl ServiceSchema {
    /// Creates a new named service schema with an empty request and response.
    #[must_use]
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            request: None,
            response: None,
        }
    }

    /// Adds request schema information.
    #[must_use]
    pub fn with_request(mut self, encoding: impl Into<String>, schema: Schema) -> Self {
        self.request = Some(MessageSchema {
            encoding: encoding.into(),
            schema,
        });
        self
    }

    /// Adds response schema information.
    #[must_use]
    pub fn with_response(mut self, encoding: impl Into<String>, schema: Schema) -> Self {
        self.response = Some(MessageSchema {
            encoding: encoding.into(),
            schema,
        });
        self
    }

    /// Returns the name of the schema.
    pub fn name(&self) -> &str {
        &self.name
    }

    /// Returns the request schema.
    pub(crate) fn request(&self) -> Option<&MessageSchema> {
        self.request.as_ref()
    }

    /// Returns the response schema.
    pub(crate) fn response(&self) -> Option<&MessageSchema> {
        self.response.as_ref()
    }
}
