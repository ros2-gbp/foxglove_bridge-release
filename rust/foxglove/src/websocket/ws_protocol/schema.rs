//! Schema types.

use std::borrow::Cow;

use base64::prelude::*;

/// An error that occurs while encoding schema data.
#[derive(Debug, thiserror::Error)]
pub enum EncodeError {
    /// The schema data is not valid UTF-8.
    #[error("Schema data is not valid UTF-8: {0}")]
    Utf8(#[from] std::str::Utf8Error),
    /// Missing schema or schema_encoding.
    #[error("Missing schema or schema_encoding")]
    MissingSchema,
}
impl From<std::string::FromUtf8Error> for EncodeError {
    fn from(err: std::string::FromUtf8Error) -> Self {
        EncodeError::Utf8(err.utf8_error())
    }
}

/// An error that occurs while decoding schema data.
#[derive(Debug, thiserror::Error)]
pub enum DecodeError {
    /// The schema data is not valid base64.
    #[error("Schema data is not valid base64: {0}")]
    Base64(#[from] base64::DecodeError),
    /// Missing schema or schema_encoding.
    #[error("Missing schema or schema_encoding")]
    MissingSchema,
}

/// Returns true if a schema is required for this message encoding.
pub(crate) fn is_schema_required(message_encoding: &str) -> bool {
    message_encoding == "flatbuffer"
        || message_encoding == "protobuf"
        || message_encoding == "ros1"
        || message_encoding == "cdr"
}

/// Returns true if the given `schema_encoding` is one of the types that is known to require binary
/// schema data (i.e. `protobuf` and `flatbuffer`). These require base64-encoding/decoding to be
/// sent via JSON messages on the websocket connection.
pub(crate) fn is_known_binary_schema_encoding(schema_encoding: impl AsRef<str>) -> bool {
    let schema_encoding = schema_encoding.as_ref();
    schema_encoding == "protobuf" || schema_encoding == "flatbuffer"
}

/// Encodes schema data, based on the schema encoding.
///
/// For binary encodings, the schema data is base64-encoded. For other encodings, the schema must
/// be valid UTF-8, or this function will return an error.
pub(crate) fn encode_schema_data<'a>(
    schema_encoding: &str,
    data: Cow<'a, [u8]>,
) -> Result<Cow<'a, str>, EncodeError> {
    if is_known_binary_schema_encoding(schema_encoding) {
        Ok(Cow::Owned(BASE64_STANDARD.encode(data)))
    } else {
        Ok(match data {
            Cow::Owned(data) => Cow::Owned(String::from_utf8(data)?),
            Cow::Borrowed(data) => Cow::Borrowed(std::str::from_utf8(data)?),
        })
    }
}

/// Decodes schema data, based on the schema encoding.
///
/// For binary encodings, the schema data is base64-encoded. For other encodings, the schema must
/// be valid UTF-8, or this function will return an error.
pub(crate) fn decode_schema_data(
    schema_encoding: &str,
    data: &str,
) -> Result<Vec<u8>, DecodeError> {
    if is_known_binary_schema_encoding(schema_encoding) {
        Ok(BASE64_STANDARD.decode(data)?)
    } else {
        Ok(data.as_bytes().to_vec())
    }
}

/// A schema for a channel or service.
#[derive(Debug, Clone)]
pub struct Schema<'a> {
    /// Schema name.
    pub name: Cow<'a, str>,
    /// Schema encoding.
    pub encoding: Cow<'a, str>,
    /// Schema data.
    pub data: Cow<'a, [u8]>,
}

impl<'a> Schema<'a> {
    /// Creates a new schema.
    pub fn new(
        name: impl Into<Cow<'a, str>>,
        encoding: impl Into<Cow<'a, str>>,
        data: impl Into<Cow<'a, [u8]>>,
    ) -> Self {
        Self {
            name: name.into(),
            encoding: encoding.into(),
            data: data.into(),
        }
    }

    /// Returns an owned version of this message.
    pub fn into_owned(self) -> Schema<'static> {
        Schema {
            name: self.name.into_owned().into(),
            encoding: self.encoding.into_owned().into(),
            data: self.data.into_owned().into(),
        }
    }
}
