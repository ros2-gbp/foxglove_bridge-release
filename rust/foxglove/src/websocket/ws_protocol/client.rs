//! Client messages.

use bytes::Buf;
use serde::Deserialize;

use crate::websocket::ws_protocol::{BinaryMessage, ParseError};

pub mod advertise;
mod fetch_asset;
mod get_parameters;
mod message_data;
mod service_call_request;
mod set_parameters;
pub mod subscribe;
mod subscribe_connection_graph;
mod subscribe_parameter_updates;
mod unadvertise;
mod unsubscribe;
mod unsubscribe_connection_graph;
mod unsubscribe_parameter_updates;

pub use advertise::Advertise;
pub use fetch_asset::FetchAsset;
pub use get_parameters::GetParameters;
pub use message_data::MessageData;
pub use service_call_request::ServiceCallRequest;
pub use set_parameters::SetParameters;
pub use subscribe::{Subscribe, Subscription};
pub use subscribe_connection_graph::SubscribeConnectionGraph;
pub use subscribe_parameter_updates::SubscribeParameterUpdates;
pub use unadvertise::Unadvertise;
pub use unsubscribe::Unsubscribe;
pub use unsubscribe_connection_graph::UnsubscribeConnectionGraph;
pub use unsubscribe_parameter_updates::UnsubscribeParameterUpdates;

/// A representation of a client message useful for deserializing.
#[derive(Debug, Clone, PartialEq)]
#[allow(missing_docs)]
pub enum ClientMessage<'a> {
    Subscribe(Subscribe),
    Unsubscribe(Unsubscribe),
    Advertise(Advertise<'a>),
    Unadvertise(Unadvertise),
    MessageData(MessageData<'a>),
    GetParameters(GetParameters),
    SetParameters(SetParameters),
    SubscribeParameterUpdates(SubscribeParameterUpdates),
    UnsubscribeParameterUpdates(UnsubscribeParameterUpdates),
    ServiceCallRequest(ServiceCallRequest<'a>),
    SubscribeConnectionGraph,
    UnsubscribeConnectionGraph,
    FetchAsset(FetchAsset),
}

impl<'a> ClientMessage<'a> {
    /// Parses a client message from JSON.
    pub fn parse_json(json: &'a str) -> Result<Self, ParseError> {
        let msg = serde_json::from_str::<JsonMessage>(json)?;
        Ok(msg.into())
    }

    /// Parses a client message from a binary buffer.
    pub fn parse_binary(mut data: &'a [u8]) -> Result<Self, ParseError> {
        if data.is_empty() {
            Err(ParseError::EmptyBinaryMessage)
        } else {
            let opcode = data.get_u8();
            match BinaryOpcode::from_repr(opcode) {
                Some(BinaryOpcode::MessageData) => {
                    MessageData::parse_binary(data).map(ClientMessage::MessageData)
                }
                Some(BinaryOpcode::ServiceCallRequest) => {
                    ServiceCallRequest::parse_binary(data).map(ClientMessage::ServiceCallRequest)
                }
                None => Err(ParseError::InvalidOpcode(opcode)),
            }
        }
    }

    /// Returns a client message with a static lifetime.
    #[allow(dead_code)]
    pub fn into_owned(self) -> ClientMessage<'static> {
        match self {
            ClientMessage::Subscribe(m) => ClientMessage::Subscribe(m),
            ClientMessage::Unsubscribe(m) => ClientMessage::Unsubscribe(m),
            ClientMessage::Advertise(m) => ClientMessage::Advertise(m.into_owned()),
            ClientMessage::Unadvertise(m) => ClientMessage::Unadvertise(m),
            ClientMessage::MessageData(m) => ClientMessage::MessageData(m.into_owned()),
            ClientMessage::GetParameters(m) => ClientMessage::GetParameters(m),
            ClientMessage::SetParameters(m) => ClientMessage::SetParameters(m),
            ClientMessage::SubscribeParameterUpdates(m) => {
                ClientMessage::SubscribeParameterUpdates(m)
            }
            ClientMessage::UnsubscribeParameterUpdates(m) => {
                ClientMessage::UnsubscribeParameterUpdates(m)
            }
            ClientMessage::ServiceCallRequest(m) => {
                ClientMessage::ServiceCallRequest(m.into_owned())
            }
            ClientMessage::SubscribeConnectionGraph => ClientMessage::SubscribeConnectionGraph,
            ClientMessage::UnsubscribeConnectionGraph => ClientMessage::UnsubscribeConnectionGraph,
            ClientMessage::FetchAsset(m) => ClientMessage::FetchAsset(m),
        }
    }
}

#[derive(Deserialize)]
#[serde(tag = "op", rename_all = "camelCase")]
enum JsonMessage<'a> {
    Subscribe(Subscribe),
    Unsubscribe(Unsubscribe),
    #[serde(borrow)]
    Advertise(Advertise<'a>),
    Unadvertise(Unadvertise),
    GetParameters(GetParameters),
    SetParameters(SetParameters),
    SubscribeParameterUpdates(SubscribeParameterUpdates),
    UnsubscribeParameterUpdates(UnsubscribeParameterUpdates),
    SubscribeConnectionGraph,
    UnsubscribeConnectionGraph,
    FetchAsset(FetchAsset),
}

impl<'a> From<JsonMessage<'a>> for ClientMessage<'a> {
    fn from(m: JsonMessage<'a>) -> Self {
        match m {
            JsonMessage::Subscribe(m) => Self::Subscribe(m),
            JsonMessage::Unsubscribe(m) => Self::Unsubscribe(m),
            JsonMessage::Advertise(m) => Self::Advertise(m),
            JsonMessage::Unadvertise(m) => Self::Unadvertise(m),
            JsonMessage::GetParameters(m) => Self::GetParameters(m),
            JsonMessage::SetParameters(m) => Self::SetParameters(m),
            JsonMessage::SubscribeParameterUpdates(m) => Self::SubscribeParameterUpdates(m),
            JsonMessage::UnsubscribeParameterUpdates(m) => Self::UnsubscribeParameterUpdates(m),
            JsonMessage::SubscribeConnectionGraph => Self::SubscribeConnectionGraph,
            JsonMessage::UnsubscribeConnectionGraph => Self::UnsubscribeConnectionGraph,
            JsonMessage::FetchAsset(m) => Self::FetchAsset(m),
        }
    }
}

#[repr(u8)]
enum BinaryOpcode {
    MessageData = 1,
    ServiceCallRequest = 2,
}
impl BinaryOpcode {
    fn from_repr(value: u8) -> Option<Self> {
        match value {
            1 => Some(Self::MessageData),
            2 => Some(Self::ServiceCallRequest),
            _ => None,
        }
    }
}
