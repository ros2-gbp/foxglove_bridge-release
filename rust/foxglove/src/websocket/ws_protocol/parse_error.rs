/// An error encountered while parsing a message.
#[derive(Debug, thiserror::Error)]
#[non_exhaustive]
pub enum ParseError {
    /// Empty binary message.
    #[error("Empty binary message")]
    EmptyBinaryMessage,
    /// Unhandled message type.
    #[error("Unhandled message type")]
    UnhandledMessageType,
    /// Invalid opcode for a binary message.
    #[error("Unknown binary opcode {0}")]
    InvalidOpcode(u8),
    /// The buffer for a binary message was too short to decode.
    #[error("Buffer too short")]
    BufferTooShort,
    /// The fetch asset response contained an invalid status code.
    #[error("Invalid fetch asset status {0}")]
    InvalidFetchAssetStatus(u8),
    /// Invalid UTF-8.
    #[error(transparent)]
    Utf8(#[from] std::str::Utf8Error),
    /// Invalid JSON.
    #[error(transparent)]
    Json(#[from] serde_json::Error),
}
