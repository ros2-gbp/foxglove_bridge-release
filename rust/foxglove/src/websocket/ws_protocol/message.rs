use std::any::type_name;

use serde::Serialize;

use crate::websocket::ws_protocol::ParseError;

/// Trait for JSON-serializable messages.
pub trait JsonMessage: Serialize {
    /// Converts the message to a JSON string.
    ///
    /// This is infallible since we control the types that implement this trait
    /// and ensure they can always be serialized to JSON.
    fn to_string(&self) -> String {
        serde_json::to_string(self)
            .unwrap_or_else(|_| panic!("failed to encode {} to JSON", type_name::<Self>()))
    }
}

/// Trait for a binary message.
pub trait BinaryMessage<'a>: Sized + 'a {
    /// Parses a binary message from the provided buffer.
    ///
    /// The caller is responsible for stripping off the opcode.
    fn parse_binary(data: &'a [u8]) -> Result<Self, ParseError>;

    /// Encodes a binary message to a new mutable buffer.
    fn to_bytes(&self) -> Vec<u8>;
}
