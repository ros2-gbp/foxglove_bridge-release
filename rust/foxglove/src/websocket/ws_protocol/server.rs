//! Server messages.

use bytes::Buf;
use serde::Deserialize;

use crate::websocket::ws_protocol::{BinaryMessage, ParseError};

pub mod advertise;
pub mod advertise_services;
pub mod connection_graph_update;
pub mod fetch_asset_response;
mod message_data;
mod parameter_values;
mod remove_status;
pub mod server_info;
mod service_call_failure;
mod service_call_response;
pub mod status;
mod time;
mod unadvertise;
mod unadvertise_services;

pub use advertise::{Advertise, Channel};
pub use advertise_services::AdvertiseServices;
pub use connection_graph_update::ConnectionGraphUpdate;
pub use fetch_asset_response::FetchAssetResponse;
pub use message_data::MessageData;
pub use parameter_values::ParameterValues;
pub use remove_status::RemoveStatus;
pub use server_info::ServerInfo;
pub use service_call_failure::ServiceCallFailure;
pub use service_call_response::ServiceCallResponse;
pub use status::Status;
pub use time::Time;
pub use unadvertise::Unadvertise;
pub use unadvertise_services::UnadvertiseServices;

/// A representation of a server message useful for deserializing.
#[derive(Debug, Clone, PartialEq)]
#[allow(missing_docs)]
pub enum ServerMessage<'a> {
    ServerInfo(ServerInfo),
    Status(Status),
    RemoveStatus(RemoveStatus),
    Advertise(Advertise<'a>),
    Unadvertise(Unadvertise),
    MessageData(MessageData<'a>),
    Time(Time),
    ParameterValues(ParameterValues),
    AdvertiseServices(AdvertiseServices<'a>),
    UnadvertiseServices(UnadvertiseServices),
    ServiceCallResponse(ServiceCallResponse<'a>),
    ConnectionGraphUpdate(ConnectionGraphUpdate),
    FetchAssetResponse(FetchAssetResponse<'a>),
    ServiceCallFailure(ServiceCallFailure),
}

impl<'a> ServerMessage<'a> {
    /// Parses a server message from JSON.
    pub fn parse_json(json: &'a str) -> Result<Self, ParseError> {
        let msg = serde_json::from_str::<JsonMessage>(json)?;
        Ok(msg.into())
    }

    /// Parses a server message from a binary buffer.
    pub fn parse_binary(mut data: &'a [u8]) -> Result<Self, ParseError> {
        if data.is_empty() {
            Err(ParseError::EmptyBinaryMessage)
        } else {
            let opcode = data.get_u8();
            match BinaryOpcode::from_repr(opcode) {
                Some(BinaryOpcode::MessageData) => {
                    MessageData::parse_binary(data).map(ServerMessage::MessageData)
                }
                Some(BinaryOpcode::Time) => Time::parse_binary(data).map(ServerMessage::Time),
                Some(BinaryOpcode::ServiceCallResponse) => {
                    ServiceCallResponse::parse_binary(data).map(ServerMessage::ServiceCallResponse)
                }
                Some(BinaryOpcode::FetchAssetResponse) => {
                    FetchAssetResponse::parse_binary(data).map(ServerMessage::FetchAssetResponse)
                }
                None => Err(ParseError::InvalidOpcode(opcode)),
            }
        }
    }

    /// Returns a server message with a static lifetime.
    #[allow(dead_code)]
    pub fn into_owned(self) -> ServerMessage<'static> {
        match self {
            ServerMessage::ServerInfo(m) => ServerMessage::ServerInfo(m),
            ServerMessage::Status(m) => ServerMessage::Status(m),
            ServerMessage::RemoveStatus(m) => ServerMessage::RemoveStatus(m),
            ServerMessage::Advertise(m) => ServerMessage::Advertise(m.into_owned()),
            ServerMessage::Unadvertise(m) => ServerMessage::Unadvertise(m),
            ServerMessage::MessageData(m) => ServerMessage::MessageData(m.into_owned()),
            ServerMessage::Time(m) => ServerMessage::Time(m),
            ServerMessage::ParameterValues(m) => ServerMessage::ParameterValues(m),
            ServerMessage::AdvertiseServices(m) => ServerMessage::AdvertiseServices(m.into_owned()),
            ServerMessage::UnadvertiseServices(m) => ServerMessage::UnadvertiseServices(m),
            ServerMessage::ServiceCallResponse(m) => {
                ServerMessage::ServiceCallResponse(m.into_owned())
            }
            ServerMessage::ConnectionGraphUpdate(m) => ServerMessage::ConnectionGraphUpdate(m),
            ServerMessage::FetchAssetResponse(m) => {
                ServerMessage::FetchAssetResponse(m.into_owned())
            }
            ServerMessage::ServiceCallFailure(m) => ServerMessage::ServiceCallFailure(m),
        }
    }
}

#[derive(Deserialize)]
#[serde(tag = "op", rename_all = "camelCase")]
enum JsonMessage<'a> {
    ServerInfo(ServerInfo),
    Status(Status),
    RemoveStatus(RemoveStatus),
    #[serde(borrow)]
    Advertise(Advertise<'a>),
    Unadvertise(Unadvertise),
    ParameterValues(ParameterValues),
    #[serde(borrow)]
    AdvertiseServices(AdvertiseServices<'a>),
    UnadvertiseServices(UnadvertiseServices),
    ConnectionGraphUpdate(ConnectionGraphUpdate),
    ServiceCallFailure(ServiceCallFailure),
}

impl<'a> From<JsonMessage<'a>> for ServerMessage<'a> {
    fn from(m: JsonMessage<'a>) -> Self {
        match m {
            JsonMessage::ServerInfo(m) => Self::ServerInfo(m),
            JsonMessage::Status(m) => Self::Status(m),
            JsonMessage::RemoveStatus(m) => Self::RemoveStatus(m),
            JsonMessage::Advertise(m) => Self::Advertise(m),
            JsonMessage::Unadvertise(m) => Self::Unadvertise(m),
            JsonMessage::ParameterValues(m) => Self::ParameterValues(m),
            JsonMessage::AdvertiseServices(m) => Self::AdvertiseServices(m.into_owned()),
            JsonMessage::UnadvertiseServices(m) => Self::UnadvertiseServices(m),
            JsonMessage::ConnectionGraphUpdate(m) => Self::ConnectionGraphUpdate(m),
            JsonMessage::ServiceCallFailure(m) => Self::ServiceCallFailure(m),
        }
    }
}

#[repr(u8)]
enum BinaryOpcode {
    MessageData = 1,
    Time = 2,
    ServiceCallResponse = 3,
    FetchAssetResponse = 4,
}
impl BinaryOpcode {
    fn from_repr(value: u8) -> Option<Self> {
        match value {
            1 => Some(Self::MessageData),
            2 => Some(Self::Time),
            3 => Some(Self::ServiceCallResponse),
            4 => Some(Self::FetchAssetResponse),
            _ => None,
        }
    }
}
