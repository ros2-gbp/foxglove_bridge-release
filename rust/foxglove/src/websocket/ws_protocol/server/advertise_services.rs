//! Advertise services message types.

use std::borrow::Cow;
use std::convert::TryFrom;

use serde::{Deserialize, Serialize};

use crate::websocket::ws_protocol::schema::{self, Schema};
use crate::websocket::ws_protocol::JsonMessage;

/// Advertise services message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#advertise-services>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "advertiseServices", rename_all = "camelCase")]
pub struct AdvertiseServices<'a> {
    /// Services.
    #[serde(borrow)]
    pub services: Vec<Service<'a>>,
}

impl<'a> AdvertiseServices<'a> {
    /// Creates a new advertise services message.
    pub fn new(services: impl IntoIterator<Item = Service<'a>>) -> Self {
        Self {
            services: services.into_iter().collect(),
        }
    }

    /// Returns an owned version of this message.
    pub fn into_owned(self) -> AdvertiseServices<'static> {
        AdvertiseServices {
            services: self.services.into_iter().map(|s| s.into_owned()).collect(),
        }
    }
}

impl JsonMessage for AdvertiseServices<'_> {}

/// A service in a [`AdvertiseServices`] message.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Service<'a> {
    /// Service ID.
    pub id: u32,
    /// Service name.
    #[serde(borrow)]
    pub name: Cow<'a, str>,
    /// Service type, which may be used to derive the request & response schema.
    #[serde(borrow)]
    pub r#type: Cow<'a, str>,
    /// Request schema.
    #[serde(borrow, skip_serializing_if = "Option::is_none")]
    pub request: Option<MessageSchema<'a>>,
    /// Request schema name.
    #[serde(borrow, skip_serializing_if = "Option::is_none")]
    pub request_schema: Option<Cow<'a, str>>,
    /// Response schema.
    #[serde(borrow, skip_serializing_if = "Option::is_none")]
    pub response: Option<MessageSchema<'a>>,
    /// Response schema name.
    #[serde(borrow, skip_serializing_if = "Option::is_none")]
    pub response_schema: Option<Cow<'a, str>>,
}

impl<'a> Service<'a> {
    /// Creates a new service advertisement.
    pub fn new(id: u32, name: impl Into<Cow<'a, str>>, r#type: impl Into<Cow<'a, str>>) -> Self {
        Self {
            id,
            name: name.into(),
            r#type: r#type.into(),
            request: None,
            request_schema: Some("".into()),
            response: None,
            response_schema: Some("".into()),
        }
    }

    /// Adds a request schema to the service advertisement.
    ///
    /// This is the preferred method for setting request schemas. Note that calling
    /// [`Service::with_request_schema`] after this method will override the schema set here.
    pub fn with_request(
        mut self,
        encoding: impl Into<Cow<'a, str>>,
        schema: Schema<'a>,
    ) -> Result<Self, schema::EncodeError> {
        let schema_data = schema::encode_schema_data(&schema.encoding, schema.data)?;
        self.request = Some(MessageSchema::new(
            encoding,
            schema.name,
            schema.encoding,
            schema_data,
        ));
        self.request_schema = None;
        Ok(self)
    }

    /// Adds a response schema to the service advertisement.
    ///
    /// This is the preferred method for setting response schemas. Note that calling
    /// [`Service::with_response_schema`] after this method will override the schema set here.
    pub fn with_response(
        mut self,
        encoding: impl Into<Cow<'a, str>>,
        schema: Schema<'a>,
    ) -> Result<Self, schema::EncodeError> {
        let schema_data = schema::encode_schema_data(&schema.encoding, schema.data)?;
        self.response = Some(MessageSchema::new(
            encoding,
            schema.name,
            schema.encoding,
            schema_data,
        ));
        self.response_schema = None;
        Ok(self)
    }

    /// Adds a request schema name to the service advertisement.
    ///
    /// This method is provided for backwards compatibility. Prefer using [`Service::with_request`]
    /// instead. Note that this will override any schema set by [`Service::with_request`].
    #[must_use]
    pub fn with_request_schema(mut self, schema_name: impl Into<Cow<'a, str>>) -> Self {
        self.request = None;
        self.request_schema = Some(schema_name.into());
        self
    }

    /// Adds a response schema name to the service advertisement.
    ///
    /// This method is provided for backwards compatibility. Prefer using
    /// [`Service::with_response`] instead. Note that this will override any schema set by
    /// [`Service::with_response`].
    #[must_use]
    pub fn with_response_schema(mut self, schema_name: impl Into<Cow<'a, str>>) -> Self {
        self.response = None;
        self.response_schema = Some(schema_name.into());
        self
    }

    /// Returns an owned version of this service.
    pub fn into_owned(self) -> Service<'static> {
        Service {
            id: self.id,
            name: self.name.into_owned().into(),
            r#type: self.r#type.into_owned().into(),
            request: self.request.map(|r| r.into_owned()),
            request_schema: self.request_schema.map(|r| r.into_owned().into()),
            response: self.response.map(|r| r.into_owned()),
            response_schema: self.response_schema.map(|r| r.into_owned().into()),
        }
    }
}

/// A message schema for a [`Service`].
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MessageSchema<'a> {
    /// Message encoding.
    #[serde(borrow)]
    pub encoding: Cow<'a, str>,
    /// Schema name.
    #[serde(borrow)]
    pub schema_name: Cow<'a, str>,
    /// Schema encoding.
    #[serde(borrow)]
    pub schema_encoding: Cow<'a, str>,
    /// Schema data.
    ///
    /// This is the protocol-encoded form. You can use the [`MessageSchema::schema`] method to
    /// decode it.
    #[serde(borrow)]
    pub schema: Cow<'a, str>,
}

impl<'a> MessageSchema<'a> {
    /// Creates a new message schema.
    pub fn new(
        encoding: impl Into<Cow<'a, str>>,
        schema_name: impl Into<Cow<'a, str>>,
        schema_encoding: impl Into<Cow<'a, str>>,
        schema: impl Into<Cow<'a, str>>,
    ) -> Self {
        Self {
            encoding: encoding.into(),
            schema_name: schema_name.into(),
            schema_encoding: schema_encoding.into(),
            schema: schema.into(),
        }
    }

    /// Returns the decoded schema data.
    pub fn schema(&self) -> Result<Vec<u8>, schema::DecodeError> {
        schema::decode_schema_data(&self.schema_encoding, &self.schema)
    }

    /// Returns an owned version of this schema.
    pub fn into_owned(self) -> MessageSchema<'static> {
        MessageSchema {
            encoding: self.encoding.into_owned().into(),
            schema_name: self.schema_name.into_owned().into(),
            schema_encoding: self.schema_encoding.into_owned().into(),
            schema: Cow::Owned(self.schema.into_owned()),
        }
    }
}

impl<'a> TryFrom<MessageSchema<'a>> for Schema<'a> {
    type Error = schema::DecodeError;

    fn try_from(value: MessageSchema<'a>) -> Result<Self, Self::Error> {
        let schema = value.schema()?;
        Ok(Schema::new(
            value.schema_name,
            value.schema_encoding,
            schema,
        ))
    }
}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::server::ServerMessage;

    use super::*;

    fn message() -> AdvertiseServices<'static> {
        AdvertiseServices::new([
            // Service with both request and response schemas using the new format
            Service::new(10, "/s1", "my_type")
                .with_request(
                    "json",
                    Schema::new("request-schema", "jsonschema", br#"{"type": "object"}"#),
                )
                .unwrap()
                .with_response(
                    "json",
                    Schema::new("response-schema", "jsonschema", br#"{"type": "object"}"#),
                )
                .unwrap(),
            // Service with only request schema using the new format
            Service::new(20, "/s2", "other_type")
                .with_request(
                    "protobuf",
                    Schema::new("request-schema", "protobuf", &[0xde, 0xad, 0xbe, 0xef]),
                )
                .unwrap(),
            // Service with both request and response schemas using the old format
            Service::new(30, "/s3", "old_type")
                .with_request_schema("request-schema")
                .with_response_schema("response-schema"),
            // Service with request schema in new format and response in old format
            Service::new(40, "/s4", "mixed_type")
                .with_request(
                    "json",
                    Schema::new("request-schema", "jsonschema", br#"{"type": "object"}"#),
                )
                .unwrap()
                .with_response_schema("response-schema"),
            // Service with request schema in old format and response in new format
            Service::new(50, "/s5", "mixed_type")
                .with_request_schema("request-schema")
                .with_response(
                    "json",
                    Schema::new("response-schema", "jsonschema", br#"{"type": "object"}"#),
                )
                .unwrap(),
            // Service with schema overrides
            Service::new(60, "/s6", "override_type")
                .with_request(
                    "json",
                    Schema::new("request-schema", "jsonschema", br#"{"type": "object"}"#),
                )
                .unwrap()
                .with_request_schema("new-request-schema")
                .with_response_schema("response-schema")
                .with_response(
                    "json",
                    Schema::new(
                        "new-response-schema",
                        "jsonschema",
                        br#"{"type": "object"}"#,
                    ),
                )
                .unwrap(),
            // Service with default schemas
            Service::new(70, "/s7", "default_schemas"),
        ])
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ServerMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ServerMessage::AdvertiseServices(orig));
    }
}
