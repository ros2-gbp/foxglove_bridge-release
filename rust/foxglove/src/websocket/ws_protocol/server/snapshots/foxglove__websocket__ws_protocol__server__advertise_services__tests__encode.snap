---
source: rust/foxglove/src/websocket/ws_protocol/server/advertise_services.rs
expression: message()
---
{
  "op": "advertiseServices",
  "services": [
    {
      "id": 10,
      "name": "/s1",
      "type": "my_type",
      "request": {
        "encoding": "json",
        "schemaName": "request-schema",
        "schemaEncoding": "jsonschema",
        "schema": "{\"type\": \"object\"}"
      },
      "response": {
        "encoding": "json",
        "schemaName": "response-schema",
        "schemaEncoding": "jsonschema",
        "schema": "{\"type\": \"object\"}"
      }
    },
    {
      "id": 20,
      "name": "/s2",
      "type": "other_type",
      "request": {
        "encoding": "protobuf",
        "schemaName": "request-schema",
        "schemaEncoding": "protobuf",
        "schema": "3q2+7w=="
      },
      "responseSchema": ""
    },
    {
      "id": 30,
      "name": "/s3",
      "type": "old_type",
      "requestSchema": "request-schema",
      "responseSchema": "response-schema"
    },
    {
      "id": 40,
      "name": "/s4",
      "type": "mixed_type",
      "request": {
        "encoding": "json",
        "schemaName": "request-schema",
        "schemaEncoding": "jsonschema",
        "schema": "{\"type\": \"object\"}"
      },
      "responseSchema": "response-schema"
    },
    {
      "id": 50,
      "name": "/s5",
      "type": "mixed_type",
      "requestSchema": "request-schema",
      "response": {
        "encoding": "json",
        "schemaName": "response-schema",
        "schemaEncoding": "jsonschema",
        "schema": "{\"type\": \"object\"}"
      }
    },
    {
      "id": 60,
      "name": "/s6",
      "type": "override_type",
      "requestSchema": "new-request-schema",
      "response": {
        "encoding": "json",
        "schemaName": "new-response-schema",
        "schemaEncoding": "jsonschema",
        "schema": "{\"type\": \"object\"}"
      }
    },
    {
      "id": 70,
      "name": "/s7",
      "type": "default_schemas",
      "requestSchema": "",
      "responseSchema": ""
    }
  ]
}
